Parser/Services/HTMLParser.cs

```diff
@@ -1,21 +1,27 @@
 ﻿using Parser.Infrastructure;
 using System.Text.RegularExpressions;
+using System.Threading.Tasks;
 
 namespace Parser.Services
 {
-    public class HTMLParser : IParser
+    public class HtmlParser : IParser
     {
-        private readonly IDataProvider _provider;
-        public HTMLParser(IDataProvider provider)
+        private readonly IDataProviderAsync _provider;
+        public HtmlParser(IDataProviderAsync provider)
         {
             _provider = provider;
         }
-        public string getSiteTitle(string URL)
+        public async Task<string> GetSiteTitleAsync(string url)
         {
-            string source = _provider.getPageHTML(URL);
-            string title = Regex.Match(source, @"\<title\b[^>]*\>\s*(?<Title>[\s\S]*?)\</title\>",
-                           RegexOptions.IgnoreCase).Groups["Title"].Value;
+            string htmlCode = await _provider.GetPageHtmlAsync(url);
+            string title = Regex.Match(htmlCode, @"\<title\b[^>]*\>\s*(?<Title>[\s\S]*?)\</title\>",
+                RegexOptions.IgnoreCase).Groups["Title"].Value;
+            if (string.IsNullOrEmpty(title))
+            {
+                return "Can't find title of the page";
```

Comment:
The service of a domain level is responsible for a business logic not for a human readable response. So here is a good place for an exception, which you may intercept on a web level and handle with a user friendly text.